File: android/app/src/main/java/pro/asciende/app/plugins/InstagramStoriesPlugin.java
package pro.asciende.app.plugins;

import android.content.Intent;
import android.net.Uri;
import android.util.Base64;

import androidx.core.content.FileProvider;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;

@CapacitorPlugin(name = "InstagramStories")
public class InstagramStoriesPlugin extends Plugin {

    @PluginMethod
    public void shareSticker(PluginCall call) {
        String base64Image = call.getString("stickerImage");
        if (base64Image == null || base64Image.isEmpty()) {
            call.reject("Missing stickerImage parameter");
            return;
        }

        String appId = call.getString("appId", "pro.asciende.app");
        String backgroundTopColor = call.getString("backgroundTopColor", "#000000");
        String backgroundBottomColor = call.getString("backgroundBottomColor", "#000000");

        try {
            byte[] imageBytes = Base64.decode(base64Image, Base64.DEFAULT);

            File cacheDir = new File(getContext().getCacheDir(), "instagram_stories");
            if (!cacheDir.exists()) {
                cacheDir.mkdirs();
            }

            File stickerFile = new File(cacheDir, "sticker_" + System.currentTimeMillis() + ".png");
            FileOutputStream fos = new FileOutputStream(stickerFile);
            fos.write(imageBytes);
            fos.flush();
            fos.close();

            Uri stickerUri = FileProvider.getUriForFile(
                getContext(),
                getContext().getPackageName() + ".fileprovider",
                stickerFile
            );

            Intent intent = new Intent("com.instagram.share.ADD_TO_STORY");
            intent.setDataAndType(stickerUri, "image/png");
            intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            intent.putExtra("source_application", appId);
            intent.putExtra("interactive_asset_uri", stickerUri);
            intent.putExtra("top_background_color", backgroundTopColor);
            intent.putExtra("bottom_background_color", backgroundBottomColor);

            if (getActivity().getPackageManager().resolveActivity(intent, 0) != null) {
                getActivity().startActivityForResult(intent, 0);
                JSObject result = new JSObject();
                result.put("success", true);
                call.resolve(result);
            } else {
                call.reject("Instagram is not installed");
            }
        } catch (IOException e) {
            call.reject("Failed to write sticker image: " + e.getMessage());
        } catch (Exception e) {
            call.reject("Failed to share to Instagram Stories: " + e.getMessage());
        }
    }
}

File: ios/App/App/Plugins/InstagramStoriesPlugin/InstagramStoriesPlugin.swift
import Foundation
import Capacitor
import UIKit

@objc(InstagramStoriesPlugin)
public class InstagramStoriesPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "InstagramStoriesPlugin"
    public let jsName = "InstagramStories"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "shareSticker", returnType: CAPPluginReturnPromise)
    ]

    @objc func shareSticker(_ call: CAPPluginCall) {
        guard let base64String = call.getString("stickerImage"),
              !base64String.isEmpty else {
            call.reject("Missing stickerImage parameter")
            return
        }

        guard let imageData = Data(base64Encoded: base64String, options: .ignoreUnknownCharacters) else {
            call.reject("Invalid base64 image data")
            return
        }

        let appId = call.getString("appId") ?? "pro.asciende.app"
        let backgroundTopColor = call.getString("backgroundTopColor") ?? "#000000"
        let backgroundBottomColor = call.getString("backgroundBottomColor") ?? "#000000"

        DispatchQueue.main.async {
            guard let url = URL(string: "instagram-stories://share") else {
                call.reject("Cannot create Instagram Stories URL")
                return
            }

            guard UIApplication.shared.canOpenURL(url) else {
                call.reject("Instagram is not installed")
                return
            }

            let pasteboardItems: [[String: Any]] = [[
                "com.instagram.sharedSticker.stickerImage": imageData,
                "com.instagram.sharedSticker.backgroundTopColor": backgroundTopColor,
                "com.instagram.sharedSticker.backgroundBottomColor": backgroundBottomColor,
                "com.instagram.sharedSticker.appID": appId
            ]]

            let options: [UIPasteboard.OptionsKey: Any] = [
                .expirationDate: Date().addingTimeInterval(60 * 5)
            ]

            UIPasteboard.general.setItems(pasteboardItems, options: options)

            UIApplication.shared.open(url, options: [:]) { success in
                if success {
                    call.resolve(["success": true])
                } else {
                    call.reject("Failed to open Instagram Stories")
                }
            }
        }
    }
}
